### PR TITLE
DB-11161 Reset JUMP_MODE to READY_TO_JUMP in prepForNextRound()

### PR DESCRIPTION
**Pull request by ascend1**

### CODE REVIEW

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0

**Comment by cloudspliceci** [comment]
TEST FAILED
https://jenkins.build.splicemachine-dev.io/job/AutomatedTesting/job/spliceengine-PR/135/

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java [hunk 1]
```diff
@@ -2152,13 +2173,23 @@ private void pullOptimizableFromJoinOrder() throws StandardException{
...
     private boolean isPreparedStatement() {
         LanguageConnectionContext lcc =
             (LanguageConnectionContext) ContextService.
                 getContextOrNull(LanguageConnectionContext.CONTEXT_ID);
         if (lcc == null || lcc.getActivationCount() < 1)
             return false;
-        return lcc.getLastActivation().getPreparedStatement() != null;
+        Activation ac = lcc.getLastActivation();
+        ResultDescription rd = ac.getResultDescription();
+        return ac.getPreparedStatement() != null && rd != null &&
```
Comment (msirek) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java:2191]: Nit: you could save a couple method calls by checking ac.getPreparedStatement() == null first, and returning right away if that is true.

File: splice_machine/src/test/java/com/splicemachine/derby/impl/sql/compile/JoinOrderJumpModeIT.java [hunk 2]
```diff
@@ -0,0 +1,262 @@
...
+    @ClassRule
+    public static TestRule chain = RuleChain.outerRule(spliceClassWatcher)
+            .around(spliceSchemaWatcher);
+    @Rule
+    public SpliceWatcher methodWatcher = new SpliceWatcher(CLASS_NAME);
+
+    public static void createData(Connection conn, String schemaName) throws Exception {
+
+        new TableCreator(conn)
+                .withCreate("CREATE TABLE \"TGSCHFAKT\" (\n" +
```
Comment (msirek) [inline splice_machine/src/test/java/com/splicemachine/derby/impl/sql/compile/JoinOrderJumpModeIT.java:33]: We should change the column names or table names if this DDL is coming from a customer.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java [hunk 3]
```diff
@@ -2152,13 +2173,23 @@ private void pullOptimizableFromJoinOrder() throws StandardException{
...
+     */
     private boolean isPreparedStatement() {
         LanguageConnectionContext lcc =
             (LanguageConnectionContext) ContextService.
                 getContextOrNull(LanguageConnectionContext.CONTEXT_ID);
         if (lcc == null || lcc.getActivationCount() < 1)
             return false;
-        return lcc.getLastActivation().getPreparedStatement() != null;
+        Activation ac = lcc.getLastActivation();
+        ResultDescription rd = ac.getResultDescription();
```
Comment (msirek) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java:2190]: I don't see the assertion in BaseActivation.getResultDescription removed.  Was this change not required after all?

**Review by msirek — APPROVED** [review]

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java (hunk 3 above)
Comment (ascend1) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java:2190]: It is removed. Please check `BaseActivation.java` in this change.

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2296/

**Comment by cloudspliceci** [comment]
TEST FAILED
https://jenkins.build.splicemachine-dev.io/job/AutomatedTesting/job/spliceengine-PR/138/

File: splice_machine/src/test/java/com/splicemachine/derby/impl/sql/compile/JoinOrderJumpModeIT.java (hunk 2 above)
Comment (ascend1) [reply splice_machine/src/test/java/com/splicemachine/derby/impl/sql/compile/JoinOrderJumpModeIT.java:33]: Yes, you're right. I renamed them in the latest commit.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java (hunk 1 above)
Comment (ascend1) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java:2191]: I think I get the point of this comment and the other about assertion is not removed. There are two `ResultDescription`s that we can get:
```
ac.getResultDescription();
```
and
```
ac.getPreparedStatement().getResultDescription();
```
At a quick glance, they return different objects. One is a member in `BaseActivation` and the other is a member in `GenericPreparedStatement`. Not sure if they eventually point to the same thing, though.

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java [hunk 4]
```diff
@@ -2152,13 +2173,23 @@ private void pullOptimizableFromJoinOrder() throws StandardException{
...
+        return ac.getPreparedStatement() != null && rd != null &&
+                !"Explain".equals(rd.getStatementType());
```
Comment (ipraznik-splice) [inline db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java:2192]: I see also many other places with if statements to compare type against string values, like "explain", "sparkexplain", "EXPLAIN", etc. Maybe it makes sense to move all these values as constants to the common class to have one place with all possible values.

**Review by ipraznik-splice — APPROVED** [review]

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2334/

File: db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java (hunk 3 above)
Comment (msirek) [reply db-engine/src/main/java/com/splicemachine/db/impl/sql/compile/OptimizerImpl.java:2190]: OK, thanks.

**Comment by ascend1** [comment]
The failing test indicates that `explain` on `update` statement with hint on the target table will be different than the actual plan. That means`isPreparedStatement()` is still risky. @msirek Why do we need to give the update statement 2 seconds to let it take the hint?

**Review by carolp-503 — APPROVED** [review]
+1

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0

**Comment by spliceci** [comment]
TEST FAILED
http://nexus.splicemachine.com:8080/job/spliceengine-PR/job/Splice-PR-Branch/2351/

**Comment by ascend1** [comment]
jenkins please test branch @cdh6.3.0,skipTestsLongerThan2Minutes

**Comment by spliceci** [comment]
TEST SUCCEEDED +1

**Pull request merged**